src/pdf_parser/pdf_utils/ocr.py

```diff
@@ -1,11 +1,329 @@
+
+    Returns:
+        Tuple of (x1, y1, x2, y2)
+    """
+    x1, y1, x2, y2 = (
+        google_coords[0].x,
+        google_coords[0].y,
+        google_coords[2].x,
+        google_coords[2].y,
+    )
+    return x1, y1, x2, y2
+
+
+def combine_google_lp(
+    image,
+    google_layout: List[GoogleTextSegment],
+    lp_layout: Layout,
+    threshold: float = 0.9,
+    top_exclude: float = 0.1,
+    bottom_exclude: float = 0.1,
+):
+    """
+    Combine google layout with layoutparser layout.
+
+    - Replaces layoutparser objects with objects recognised by google API + their text if they overlap sufficiently.
+    - Does not include google objects if they seem to be not part of the main text (e.g. headers, footers, etc.). We
+    use top exclude and bottom exclude to ascertain this.
+
+
+    Args:
+        image: The image the layout is based on.
+        google_layout: List of GoogleTextSegment objects inferred from the structure returned by the Google OCR API.
+        lp_layout: Layout object
+        threshold: Threshold for overlap between google and layoutparser objects. If the overlap is larger than
+            threshold, the layoutparser object is replaced by the google object.
+        top_exclude: Exclude objects from the top of the page if they are above this fraction of the page.
+        bottom_exclude: Exclude objects from the bottom of the page if they are below this fraction of the page.
+
+    Returns:
+        Layout object with google objects replacing layoutparser objects if they overlap sufficiently.
+    """
+    shapely_google = [google_vertex_to_shapely(b.coordinates) for b in google_layout]
+    shapely_layout = [lp_coords_to_shapely_polygon(b.coordinates) for b in lp_layout]
+    # for every block in the google layout, find the fraction of the block that is covered by the layoutparser layout
+    dd_intersection_over_union = defaultdict(list)
+    for ix_goog, google_block in enumerate(shapely_google):
+        for ix_lp, lp_block in enumerate(shapely_layout):
+            # Find the intersection over union of the two blocks.
+            intersection = google_block.intersection(lp_block).area
+            union = google_block.union(lp_block).area
+            dd_intersection_over_union[ix_goog].append(intersection / union)
+
+    # If a google block is covered by a layoutparser block by more than 0.9, we can assume that google OCR has
+    # identified the same block as layoutparser + heuristics and we can use the text from the google block.
+    # Filter the default dict for these cases
+    equivalent_block_mapping = {
+        k: v.index(max(v))
+        for k, v in dd_intersection_over_union.items()
+        if max(v) > threshold
+    }
+
+    # New blocks to add to the layoutparser layout
+    blocks_google_only = {
+        k: v for k, v in dd_intersection_over_union.items() if max(v) == 0.0
+    }
+
+    # use the mapping to replace the text of the layoutparser block with the text of the google block
```

Comment:
Could add this as a class method to a layout parser block type object?

Reply:
eh, not that keen on this as it makes the line between google blocks and layout parser blocks fuzzy.

Reply:
Fair enough I would at least have it as a function with a test however and not nested in this long block of code:


            def replace_block_text(replace_block: GoogleBlock, to_be_replace_block: LayoutBlock) -> LayoutBlock:
                    """"use the mapping to replace the text of the layoutparser block with the text of the google block""""
                    
                    return ... 

            def test_replace_block_text....